### DISCUSSION TITLE
Self-referential Composite foreign key overlapping primary key

### DISCUSSION THREAD

**Discussion by asbuch** [sqlalchemy/sqlalchemy#11992] [Usage Questions]
### Describe the bug

I think a composite foreign key should be considered NULL if any of its components is NULL.
SQLAlchemy's behavior is sometimes consistent with this, sometimes not.

Consistent:  JOIN operations use equality "parent_idx = idx", which evaluates to False on NULL values.
Inconsistent:  Reading attributes can result in "idx IS NULL", which evaluates to True.

### SQLAlchemy Version in Use

2.0.35

### DBAPI (i.e. the database driver)

pysqlite

### Database Vendor and Major Version

3.37.2

### Python Version

3.10

### Operating system

Linux

### To Reproduce

```python
import sqlalchemy as sa
import sqlalchemy.orm

class Base(sa.orm.DeclarativeBase):
    pass

class Section(Base):
    __tablename__ = 'sections'
    year = sa.Column(sa.Integer, primary_key=True)
    idx = sa.Column(sa.Integer, primary_key=True)
    parent_idx = sa.Column(sa.Integer)

    sa.ForeignKeyConstraint((year, parent_idx), (year, idx))

    children = sa.orm.relationship(
        'Section',
        primaryjoin=sa.and_(
            year == sa.orm.remote(year),
            idx == sa.orm.remote(sa.orm.foreign(parent_idx))),
        back_populates='parent',
        post_update=True,
    )
    parent = sa.orm.relationship(
        'Section',
        primaryjoin=sa.and_(
            year == sa.orm.remote(year),
            sa.orm.foreign(parent_idx) == sa.orm.remote(idx)),
        back_populates='children',
        post_update=True,
    )

engine = sa.create_engine('sqlite+pysqlite:///:memory:', echo=True)
Base.metadata.create_all(engine)
session = sa.orm.Session(engine)

session.add(Section(year=5, idx=1, parent_idx=None))
session.commit()

s = session.get(Section, (5, 1))
print(s.parent)
# Issues "SELECT ... WHERE ... AND sections.idx IS NULL"
# Here "NULL IS NULL" evaluates to True.

query = sa.select(Section).options(sa.orm.joinedload(Section.parent))
session.scalars(query).all()
# Issues "... JOIN ... sections.parent_idx = sections_1.idx"
# Here "NULL = NULL" evaluates to False.
```

### Error

There is no error output, but the queries are not optimal.
The comments in the code indicate where inconsistent queries are issued.
In the line "print(s.parent)", it would better if no SELECT is issued.

### Additional context

Thanks for making this great package available!!!

**Comment by zzzeek** [comment]
composite foreign keys can refer to rows where some of the columns are NULL.  Here's a nonsensical example, here run on PostgreSQL which has the most strict referential integrity rules

```py
from sqlalchemy import and_
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import engine
from sqlalchemy import ForeignKeyConstraint
from sqlalchemy import UniqueConstraint
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import foreign
from sqlalchemy.orm import joinedload
from sqlalchemy.orm import query
from sqlalchemy.orm import relationship
from sqlalchemy.orm import remote
from sqlalchemy.orm import Session
from sqlalchemy.orm import session


class Base(DeclarativeBase):
    pass


class Section(Base):
    __tablename__ = "sections"
    id = Column(Integer, primary_key=True)
    year = Column(Integer)
    idx = Column(Integer)
    parent_idx = Column(Integer)

    __mapper_args__ = {"primary_key": [year, idx]}

    ForeignKeyConstraint((year, parent_idx), (year, idx))
    UniqueConstraint(year, idx)

    children = relationship(
        "Section",
        primaryjoin=and_(
            year == remote(year), idx == remote(foreign(parent_idx))
        ),
        back_populates="parent",
        post_update=True,
    )
    parent = relationship(
        "Section",
        primaryjoin=and_(
            year == remote(year), foreign(parent_idx) == remote(idx)
        ),
        back_populates="children",
        post_update=True,
    )


engine = create_engine("postgresql://scott:tiger@localhost/test", echo=True)
Base.metadata.create_all(engine)
session = Session(engine)

session.add(Section(year=5, idx=1, parent_idx=None))
session.add(Section(year=5, idx=None, parent_idx=None))
session.commit()

s = session.get(Section, (5, 1))
print(s.parent)

```

section (5,1) has a .parent that is (5, None)

```
2024-10-12 20:06:12,159 INFO sqlalchemy.engine.Engine SELECT sections.id AS sections_id, sections.year AS sections_year, sections.idx AS sections_idx, sections.parent_idx AS sections_parent_idx 
FROM sections 
WHERE sections.year = %(pk_1)s AND sections.idx IS NULL
2024-10-12 20:06:12,159 INFO sqlalchemy.engine.Engine [generated in 0.00005s] {'pk_1': 5}
<__main__.Section object at 0x7f547c6cba40>

```

I'm not creative enough to come up with some of the real world cases that are currently out there where trying to guess about the intent of foreign key constraints would be a bad idea, but they are definitely there, we've had some very weird overlapping foreign key use cases.     ill try to think of a more interesting one

**Comment by asbuch** [comment]
Thanks, this is helpful! The following iteration uses sqlite again, and uses two additional queries to load the section. The parent of the section can be None or not, depending on the query.

```
from sqlalchemy import and_
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import engine
from sqlalchemy import ForeignKeyConstraint
from sqlalchemy import UniqueConstraint
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import foreign
from sqlalchemy.orm import joinedload
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import query
from sqlalchemy.orm import relationship
from sqlalchemy.orm import remote
from sqlalchemy.orm import Session
from sqlalchemy.orm import session


class Base(DeclarativeBase):
    pass


class Section(Base):
    __tablename__ = "sections"
    id = Column(Integer, primary_key=True)
    year = Column(Integer)
    idx = Column(Integer)
    parent_idx = Column(Integer)

    __mapper_args__ = {"primary_key": [year, idx]}

    ForeignKeyConstraint((year, parent_idx), (year, idx))
    UniqueConstraint(year, idx)

    children = relationship(
        "Section",
        primaryjoin=and_(
            year == remote(year), idx == remote(foreign(parent_idx))
        ),
        back_populates="parent",
        post_update=True,
    )
    parent = relationship(
        "Section",
        primaryjoin=and_(
            year == remote(year), foreign(parent_idx) == remote(idx)
        ),
        back_populates="children",
        post_update=True,
    )


engine = create_engine('sqlite+pysqlite:///:memory:', echo=True)
Base.metadata.create_all(engine)
session = Session(engine)

session.add(Section(year=5, idx=1, parent_idx=None))
session.add(Section(year=5, idx=None, parent_idx=None))
session.commit()

s = session.get(Section, (5, 1))
assert s.parent is not None

session.expunge(s)

s = session.scalar(
    select(Section).filter_by(idx=1).options(joinedload(Section.parent)))
assert s.parent is None

session.expunge(s)

s = session.scalar(
    select(Section).filter_by(idx=1).options(selectinload(Section.parent)))
assert s.parent is None
```

**Reply by asbuch** [reply]
This example is real-life. The sections are classes at a university and have primary key (year, term, index). Some sections are grouped together and have a parent, others have no parent.

**Reply by asbuch** [reply]
If some applications need composite foreign keys with NULL components to be valid, then they should use `isnot_distinct_from()` to define the `primaryjoin`. The condition `idx == parent_idx` ought to mean the SQL equality operator, and should not be changed to `IS NULL`, IMHO. This convention would allow skipping the SELECT query on `s.parent` when a foreign key component is NULL. Are there other ways to achieve this now?

**Reply by zzzeek** [reply]
so your bigger question "can lazy loading ever skip the SELECT" the answer is ~~no~~ "not for an arbitrary primary join [edit: only for specific get-by-primary-key scenarios, which this might end up being]", because SQLAlchemy does not attempt to evaluate the meaning of a SELECT query in memory, it uses the database for that.  The WHERE condition could be literally anything, with ORs, CASE, etc., there's no generalized way SQLAlchemy could heuristically "guess" as to what kinds of WHERE clauses with what kinds of parameters ultimately mean "this SELECT will never return rows", short of re-implementing a relational database engine in SQLAlchemy itself, which is a hard line we will never cross.  Even a WHERE criteria that literally included `true = false`, that could be part of a larger set of criteria.  The database already knows how to give us this answer and SQLAlchemy always uses that.

**Reply by zzzeek** [reply]
There is supposed to be some logic that prevents a lazyload from emitting when we are loading by primary key, as is the case here, and the primary key is NULL.    There is also logic for the mapper to consider a partially-NULL key to be NULL as well.   I'm looking to see when that logic kicks in and what is preventing it from happening here.

**Reply by zzzeek** [reply]
here's a patch that allows that:

```diff
diff --git a/lib/sqlalchemy/orm/strategies.py b/lib/sqlalchemy/orm/strategies.py
index 3f947a8d7..d228f9702 100644
--- a/lib/sqlalchemy/orm/strategies.py
+++ b/lib/sqlalchemy/orm/strategies.py
@@ -936,8 +936,12 @@ class LazyLoader(
             elif LoaderCallableStatus.NEVER_SET in primary_key_identity:
                 return LoaderCallableStatus.NEVER_SET
 
-            if _none_set.issuperset(primary_key_identity):
-                return None
+            if not self.mapper.allow_partial_pks:
+                if _none_set.intersection(primary_key_identity):
+                    return None
+            else:
+                if _none_set.issuperset(primary_key_identity):
+                    return None
 
             if (
                 self.key in state.dict

```

if you set up your mapping as:

```py
class Section(Base):
    __tablename__ = "sections"

    # ...

    __mapper_args__ = {"allow_partial_pks": False}

```

I can evaluate this as an additional rule to add.  #11995

**Reply by asbuch** [reply]
Your patch works great for my original example, thanks!!!

It appears to interfere with the updated example using:

`__mapper_args__ = {"primary_key": [year, idx]}`

That example gives a `FlushError` when `"allow_partial_pks": False` is added. This is not a concern for me since I don't use NULL primary keys. I hope some version of your patch will find its way to the official version. 

Also, I think `selectinload` is already disregarding partial keys. For example (after importing the third code example):

```
session.add(Section(year=6, idx=2, parent_idx=None))
session.add(Section(year=7, idx=3, parent_idx=99))
session.commit()
session.reset()

session.scalar(
    select(Section).filter_by(idx=2).options(selectinload(Section.parent)))
# Issues one SELECT query

session.scalar(
    select(Section).filter_by(idx=3).options(selectinload(Section.parent)))
# Issues two SELECT queries
```

**Reply by asbuch** [reply]
Probably the `FlushError` was intended, since `"allow_partial_pks": False` applies to the primary key, not to the relationship?

I think the documentation ought to say that `"allow_partial_pks": False` is likely to become the default in the future, since in almost all cases the primary key is defined as `NOT NULL` in the database, and partial keys don't work well with `joinedload` and `selectinload` anyway.
